sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -752,9 +752,30 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
       expressions.exists(_.exists(_.semanticEquals(expr)))
     }
 
+    def checkOuterReferenceUtil(attrs: Seq[Attribute], outer: OuterReference): Unit = {
+      if (!attrs.exists(_.exprId == outer.exprId)) {
+        failAnalysis("outer attribute not found")
+      }
+    }
+
+    def checkOuterReference(p: LogicalPlan, expr: SubqueryExpression): Unit = p match {
+      case f: Filter =>
```

Comment:
why do we only apply the check for Filter?

Reply:
Filter is the case where we saw this issue (Filter contains a exists or not exists and a subplan as its child).

SQL UDF is a case that is not filter but also contains outer reference. That is a case that this fix is not aim at.

Reply:
Generally speaking if we want to improve outer reference check in general, that will beyond SPARK-39144 and worth a specific JIRA to track.